commit edac64cb6cbacc5650cd25f08a4cc0316373af37
Author: egorpon <ponomarenkoegor36@gmail.com>
Date:   Wed Apr 8 19:53:52 2026 +0300

    test: add tests for payment views

diff --git a/account/tests/test_views.py b/account/tests/test_views.py
index d9fd53e..8463ffa 100644
--- a/account/tests/test_views.py
+++ b/account/tests/test_views.py
@@ -10,7 +10,6 @@ from ..token import email_token_generator
 
 from cart.factory import UserFactory
 
-import re
 
 from payment.models import ShippingAddress
 # Create your tests here.
@@ -49,11 +48,10 @@ class AccountViewTest(TestCase):
 
         self.assertIn(expected_uid, email.body)
 
-        match = re.search(r'http://[^\s]+', email.body)
-        url = match.group(0)
-        response = self.client.get(url)
-        user.refresh_from_db()
-        self.assertTrue(user.is_active)
+        token = email_token_generator.make_token(user)
+
+        self.assertIn(token, email.body)
+        self.assertTrue(email_token_generator.check_token(user, token))
 
     def test_register_user_with_invalid_credentials_returns_form_errors(self):
         response = self.client.post(
@@ -206,24 +204,6 @@ class AccountViewTest(TestCase):
 
         self.assertEqual(shipping.city, "Praga")
 
-    def test_manage_shipping_create_new(self):
-        self.client.force_login(self.user)
-
-        data = {
-            "full_name": "tester testerovich",
-            "email": "test@gmail.com",
-            "city": "Praga",
-            "address1": "Praga, St. Praga",
-        }
-
-        response = self.client.post(reverse("manage_shipping"), data=data)
-
-        self.assertRedirects(response, reverse("manage_shipping"))
-
-        shipping = ShippingAddress.objects.get(user=self.user)
-
-        self.assertEqual(shipping.city, "Praga")
-
     def test_manage_shipping_update_existing(self):
         self.client.force_login(self.user)
 
diff --git a/cart/templates/cart/cart-summary.html b/cart/templates/cart/cart-summary.html
index 24d8e32..7d74f38 100644
--- a/cart/templates/cart/cart-summary.html
+++ b/cart/templates/cart/cart-summary.html
@@ -72,7 +72,9 @@
             <div id="new-total" class="d-none h3 fw-bold text-danger"></div>
           {% endif %}
         </div>
-        <a role="button" class="btn btn-outline-primary" href="{% url 'checkout' %}">Proceed to checkout</a>
+        {% if cart %}
+          <a role="button" class="btn btn-outline-primary" href="{% url 'checkout' %}">Proceed to checkout</a>
+        {% endif %}
       </div>
     </div>
   </section>
diff --git a/cart/tests/test_services.py b/cart/tests/test_services.py
index ea89abc..155ff3f 100644
--- a/cart/tests/test_services.py
+++ b/cart/tests/test_services.py
@@ -81,13 +81,13 @@ class CartServiceTest(TestCase):
         for item in self.service:
             self.assertEqual(item["product"], self.product)
             self.assertEqual(item["quantity"], 2)
-            self.assertEqual(item["total"], round(Decimal(self.product.price * 2), 2))
+            self.assertEqual(item["total"], Decimal('25.98'))
 
     def test_get_total_calculation(self):
         self.service.add(self.product, 2)
 
         self.assertEqual(
-            self.service.get_total(), round(Decimal(self.product.price * 2), 2)
+            self.service.get_total(), Decimal('25.98')
         )
 
     def test_get_item_total_calculation(self):
diff --git a/coupon/tests/test_views.py b/coupon/tests/test_views.py
index 421b1cb..6734cc6 100644
--- a/coupon/tests/test_views.py
+++ b/coupon/tests/test_views.py
@@ -4,7 +4,9 @@ from django.utils import timezone
 from datetime import timedelta
 from django.urls import reverse
 
-from cart.factory import CartFactory, CartItemFactory, ProductFactory, CategoryFactory
+from cart.factory import UserFactory, CartFactory, CartItemFactory, ProductFactory, CategoryFactory
+
+from cart.models import Cart
 
 from decimal import Decimal
 
@@ -83,6 +85,10 @@ class CouponViewTest(TestCase):
             reverse("apply_coupon"), {"code": "SAVE12", "action": "post"}
         )
 
+        cart = Cart.objects.filter(session_key=self.client.session.session_key).first()
+        
+        self.assertEqual(cart.coupon, self.coupon)
+
         self.assertEqual(response.status_code, 200)
 
         data = response.json()
diff --git a/payment/tests/test_views.py b/payment/tests/test_views.py
index ea5aeaf..7dd717e 100644
--- a/payment/tests/test_views.py
+++ b/payment/tests/test_views.py
@@ -1,12 +1,24 @@
 from django.test import TestCase
 
 from ..forms import ShippingForm
-from ..models import ShippingAddress
+from ..models import ShippingAddress, Order, OrderItem
 
 from django.urls import reverse
 from ..factory import ShippingAddressFactory
 
-from cart.factory import UserFactory
+from cart.factory import (
+    UserFactory,
+    CartFactory,
+    CartItemFactory,
+    CategoryFactory,
+    ProductFactory,
+)
+
+from cart.models import Cart
+
+from decimal import Decimal
+
+from django.core import mail
 
 
 # Create your tests here.
@@ -17,6 +29,14 @@ class PaymentViewTest(TestCase):
             full_name="Benjamin Netanyahu", user=self.user
         )
 
+        self.category = CategoryFactory()
+        self.product = ProductFactory(category=self.category)
+
+        self.cart = CartFactory()
+        self.cart_item = CartItemFactory(
+            cart=self.cart, product=self.product, quantity=1
+        )
+
     def test_checkout_show_saved_user_shipping_address(self):
         self.client.force_login(self.user)
 
@@ -33,3 +53,110 @@ class PaymentViewTest(TestCase):
         self.assertEqual(response.status_code, 200)
         form = response.context["form"]
         self.assertEqual(form.initial, {})
+
+    def test_payment_success_deletes_user_cart(self):
+        self.cart.user = self.user
+        self.cart.save()
+
+        self.cart.refresh_from_db()
+
+        self.client.force_login(self.user)
+
+        response = self.client.get(reverse("payment_success"))
+
+        self.assertEqual(response.status_code, 200)
+        cart = Cart.objects.filter(user=self.user)
+        self.assertFalse(cart.exists())
+
+    def test_payment_success_deletes_guest_cart(self):
+        session_key = self.client.session.session_key
+        self.cart.session_key = session_key
+        self.cart.save()
+
+        self.cart.refresh_from_db()
+
+        response = self.client.get(reverse("payment_success"))
+
+        self.assertEqual(response.status_code, 200)
+        cart = Cart.objects.filter(session_key=session_key)
+        self.assertFalse(cart.exists())
+
+    def test_complete_order_authorized_user_flow(self):
+        self.client.force_login(self.user)
+        self.cart.user = self.user
+
+        self.cart.save()
+        self.cart.refresh_from_db()
+
+        data = {
+            "action": "post",
+            "name": "Benjamin",
+            "email": "test@test.com",
+            "address1": "Main St 1",
+            "city": "Tel Aviv",
+            "zipcode": "12345",
+        }
+
+        response = self.client.post(reverse("complete_order"), data=data)
+
+        self.assertEqual(response.status_code, 200)
+
+
+        order = Order.objects.filter(full_name="Benjamin", user=self.user).first()
+        self.assertIsNotNone(order)
+        self.assertEqual(order.amount_paid, Decimal("12.99"))
+
+        expected_address = 'Main St 1\nTel Aviv\n12345'
+        self.assertEqual(order.shipping_address, expected_address)
+
+        order_item = OrderItem.objects.filter(order=order).first()
+        self.assertIsNotNone(order_item)
+        self.assertEqual(order_item.product, self.product)
+
+        self.assertEqual(response.json()["success"], True)
+
+        self.assertEqual(len(mail.outbox), 1)
+        email = mail.outbox[0]
+
+        self.assertIn("Order", email.subject)
+        self.assertIn("Total paid", email.body)
+
+    def test_complete_order_unauthorized_user_flow(self):
+        session_key = self.client.session.session_key
+        self.cart.session_key = session_key
+        self.cart.save()
+
+        data = {
+            "action": "post",
+            "name": "Benjamin",
+            "email": "test@test.com",
+            "address1": "Main St 1",
+            "city": "Tel Aviv",
+            "zipcode": "12345",
+        }
+
+        response = self.client.post(reverse("complete_order"), data=data)
+
+        self.assertEqual(response.status_code, 200)
+
+
+        order = Order.objects.filter(full_name="Benjamin", user=None).first()
+        self.assertIsNotNone(order)
+        self.assertEqual(order.amount_paid, Decimal("12.99"))
+
+        expected_address = 'Main St 1\nTel Aviv\n12345'
+        self.assertEqual(order.shipping_address, expected_address)
+
+        order_item = OrderItem.objects.filter(order=order).first()
+        self.assertIsNotNone(order_item)
+        self.assertEqual(order_item.product, self.product)
+
+        self.assertEqual(response.json()["success"], True)
+
+        self.assertEqual(len(mail.outbox), 1)
+        email = mail.outbox[0]
+
+        self.assertIn("Order", email.subject)
+        self.assertIn("Total paid", email.body)
+
+
diff --git a/payment/views.py b/payment/views.py
index a0ee57b..6040e5b 100644
--- a/payment/views.py
+++ b/payment/views.py
@@ -27,16 +27,14 @@ def checkout(request):
 def payment_success(request):
     if request.user.is_authenticated:
         user_cart = Cart.objects.filter(user=request.user).first()
-        if not user_cart:
-            return redirect("payment_failed")
+
         user_cart.delete()
 
     else:
         guest_cart = Cart.objects.filter(
             session_key=request.session.session_key
         ).first()
-        if not guest_cart:
-            return redirect("payment_failed")
+
         guest_cart.delete()
 
     return render(request, "payment/payment-success.html")
